sysbuild with custom board compilation failed to find the board

**Describe the bug**
During a build with the **_sysbuild_** option the option to use a specific path for a custom board is failed.

Please also mention any information which could help others to understand
the problem you're facing:
 - target nrf52840dk_nrf52840
 - project hello_world from zephyr/samples
 - using the option --sysbuild with the option -DBOARD_ROOT then 
 - Dhello_world_BOARD_ROOT, -Dhello_world_BOARD_ROOT_CONFIG, -Dhello_world_CONFIG_BOARD_ROOT

**To Reproduce**
Steps to reproduce the behavior:
1.  west build -c -p auto -b plank --sysbuild zephyr/samples/hello_world/ -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DBOARD_ROOT=/../path_of_the_custom_boards
2. CMake Error at /../zephyrproject/zephyr/cmake/modules/boards.cmake:163 (message):
  Invalid BOARD; see above.

**Expected behavior**
The build is successful for the custom board

**Impact**
It is not possible to use the sysbuild and so all the domain mechanism if the boards is not describes into the folder zephyr/

**Logs and console output**

```
 west build -c -p auto -b plank -d build/hello_world --sysbuild zephyr/samples/hello_world/ -- -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DBOARD_ROOT=/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/test/
-- west build: generating a build system
Loading Zephyr module(s) (Zephyr base): extensions sysbuild_extensions python west root zephyr_module boards shields sysbuild_kconfig
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.6") found components: Interpreter 
-- Cache files will be written to: /home/viny/.cache/zephyr
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
-- Board: plank
Parsing /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/share/sysbuild/Kconfig
Loaded configuration '/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/empty.conf'
Merged configuration '/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/empty.conf'
Merged configuration '/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/zephyr/misc/generated/extra_kconfig_options.conf'
Merged configuration '/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/empty.conf'
Configuration saved to '/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/zephyr/.config'
Kconfig header saved to '/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/autoconf.h'
-- 
   *********************************
   * Running CMake for hello_world *
   *********************************

Loading Zephyr default modules (Zephyr base).
-- Application: /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/samples/hello_world
-- CMake version: 3.22.1
-- Found Python3: /usr/bin/python3.10 (found suitable exact version "3.10.6") found components: Interpreter 
-- Cache files will be written to: /home/viny/.cache/zephyr
-- Zephyr version: 3.2.99 (/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr)
-- Found west (found suitable version "0.14.0", minimum required is "0.7.1")
-- Board: plank
No board named 'plank' found.

Please choose one of the following boards:

arc:
  em_starterkit
  ...
CMake Error at /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/cmake/modules/boards.cmake:163 (message):
  Invalid BOARD; see above.
Call Stack (most recent call first):
  /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/cmake/modules/zephyr_default.cmake:113 (include)
  /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:5 (find_package)


-- Configuring incomplete, errors occurred!
CMake Error at cmake/modules/sysbuild_extensions.cmake:243 (message):
  CMake configure failed for Zephyr project: hello_world

  Location:
  /home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/samples/hello_world
Call Stack (most recent call first):
  CMakeLists.txt:59 (ExternalZephyrProject_Add)


-- Configuring incomplete, errors occurred!
See also "/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world/CMakeFiles/CMakeOutput.log".
FATAL ERROR: command exited with status 1: /usr/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/build/hello_world -GNinja -DBOARD=plank -DSB_CONFIG_BOOTLOADER_MCUBOOT=y -DBOARD_ROOT=/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/test/ -S/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/share/sysbuild -DAPP_DIR:PATH=/home/viny/Workspace/01-Repository/zephyr-3.2.0/zephyrproject/zephyr/samples/hello_world
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain : zephyr 3.2.99, gcc-arm-none-eabi-10-2020-q4-major-x86_64-linux.tar.bz2, zephyr-sdk-0.15.2_linux-x86_64.tar.gz

**Additional context**

--- a/cmake/modules/extensions.cmake
+++ b/cmake/modules/extensions.cmake
@@ -2467,10 +2467,10 @@ function(zephyr_list transform list_var action)
 endfunction()
 
 # Usage:
-#   zephyr_get(<variable>)
-#   zephyr_get(<variable> SYSBUILD [LOCAL|GLOBAL])
+#   zephyr_get(<variable> [MERGE] [SYSBUILD [LOCAL|GLOBAL]])
 #
-# Return the value of <variable> as local scoped variable of same name.
+# Return the value of <variable> as local scoped variable of same name. If MERGE
+# is supplied, will return a list of found items.
 #
 # zephyr_get() is a common function to provide a uniform way of supporting
 # build settings that can be set from sysbuild, CMakeLists.txt, CMake cache, or
@@ -2484,6 +2484,8 @@ endfunction()
 #   - blinky_BOARD is considered a local sysbuild cache variable only for the
 #     blinky image.
 #   If no sysbuild scope is specified, GLOBAL is assumed.
+#   If using MERGE then SYSBUILD GLOBAL will get both the local and global
+#   sysbuild scope variables (in that order, if both exist).
 # - CMake cache, set by `-D<var>=<value>` or `set(<var> <val> CACHE ...)
 # - Environment
 # - Locally in CMakeLists.txt before 'find_package(Zephyr)'
@@ -2493,7 +2495,7 @@ endfunction()
 # using `-DZEPHYR_TOOLCHAIN_VARIANT=<val>`, then the value from the cache is
 # returned.
 function(zephyr_get variable)
-  cmake_parse_arguments(GET_VAR "" "SYSBUILD" "" ${ARGN})
+  cmake_parse_arguments(GET_VAR "MERGE" "SYSBUILD" "" ${ARGN})
 
   if(DEFINED GET_VAR_SYSBUILD)
     if(NOT (${GET_VAR_SYSBUILD} STREQUAL "GLOBAL" OR
@@ -2505,6 +2507,13 @@ function(zephyr_get variable)
     set(GET_VAR_SYSBUILD "GLOBAL")
   endif()
 
+  if(GET_VAR_MERGE)
+    # Clear variable before appending items in MERGE mode
+    set(${variable})
+  endif()
+
+  set(used_global false)
+
   if(SYSBUILD)
     get_property(sysbuild_name TARGET sysbuild_cache PROPERTY SYSBUILD_NAME)
     get_property(sysbuild_main_app TARGET sysbuild_cache PROPERTY SYSBUILD_MAIN_APP)
@@ -2513,19 +2522,40 @@ function(zephyr_get variable)
        (${GET_VAR_SYSBUILD} STREQUAL "GLOBAL" OR sysbuild_main_app)
     )
       get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
+      set(used_global true)
     endif()
   endif()
 
   if(DEFINED sysbuild_${variable})
-    set(${variable} ${sysbuild_${variable}} PARENT_SCOPE)
-  elseif(DEFINED CACHE{${variable}})
-    set(${variable} $CACHE{${variable}} PARENT_SCOPE)
-  elseif(DEFINED ENV{${variable}})
-    set(${variable} $ENV{${variable}} PARENT_SCOPE)
-    # Set the environment variable in CMake cache, so that a build invocation
-    # triggering a CMake rerun doesn't rely on the environment variable still
-    # being available / have identical value.
-    set(${variable} $ENV{${variable}} CACHE INTERNAL "")
+    if(GET_VAR_MERGE)
+      list(APPEND ${variable} ${sysbuild_${variable}})
+    else()
+      set(${variable} ${sysbuild_${variable}} PARENT_SCOPE)
+      return()
+    endif()
+  endif()
+  if(SYSBUILD AND GET_VAR_MERGE AND NOT used_global AND ${GET_VAR_SYSBUILD} STREQUAL "GLOBAL")
+    get_property(sysbuild_${variable} TARGET sysbuild_cache PROPERTY ${variable})
+    list(APPEND ${variable} ${sysbuild_${variable}})
+  endif()
+  if(DEFINED CACHE{${variable}})
+    if(GET_VAR_MERGE)
+      list(APPEND ${variable} $CACHE{${variable}})
+    else()
+      set(${variable} $CACHE{${variable}} PARENT_SCOPE)
+      return()
+    endif()
+  endif()
+  if(DEFINED ENV{${variable}})
+    if(GET_VAR_MERGE)
+      list(APPEND ${variable} $ENV{${variable}}})
+    else()
+      set(${variable} $ENV{${variable}} PARENT_SCOPE)
+      # Set the environment variable in CMake cache, so that a build invocation
+      # triggering a CMake rerun doesn't rely on the environment variable still
+      # being available / have identical value.
+      set(${variable} $ENV{${variable}} CACHE INTERNAL "")
+    endif()
 
     if(DEFINED ${variable} AND NOT "${${variable}}" STREQUAL "$ENV{${variable}}")
       # Variable exists as a local scoped variable, defined in a CMakeLists.txt
@@ -2537,6 +2567,15 @@ function(zephyr_get variable)
                       "Local scope value (hidden): ${${variable}}\n"
       )
     endif()
+
+    if(NOT GET_VAR_MERGE)
+      return()
+    endif()
+  endif()
+
+  if(GET_VAR_MERGE)
+    list(REMOVE_DUPLICATES ${variable})
+    set(${variable} ${${variable}} PARENT_SCOPE)
   endif()
 endfunction(zephyr_get variable)
 

--- a/cmake/modules/root.cmake
+++ b/cmake/modules/root.cmake
@@ -36,6 +36,13 @@ zephyr_file(APPLICATION_ROOT ARCH_ROOT)
 # Convert paths to absolute, relative from APPLICATION_SOURCE_DIR
 zephyr_file(APPLICATION_ROOT SCA_ROOT)
 
+# Merge in variables from other sources (e.g. sysbuild)
+zephyr_get(MODULE_EXT_ROOT MERGE SYSBUILD GLOBAL)
+zephyr_get(BOARD_ROOT MERGE SYSBUILD GLOBAL)
+zephyr_get(SOC_ROOT MERGE SYSBUILD GLOBAL)
+zephyr_get(ARCH_ROOT MERGE SYSBUILD GLOBAL)
+zephyr_get(SCA_ROOT MERGE SYSBUILD GLOBAL)
+
 if(unittest IN_LIST Zephyr_FIND_COMPONENTS)
   # Zephyr used in unittest mode, use dedicated unittest root.
   set(BOARD_ROOT ${ZEPHYR_BASE}/subsys/testsuite)

--- a/doc/releases/release-notes-3.4.rst
+++ b/doc/releases/release-notes-3.4.rst
@@ -323,6 +323,15 @@ Build system and infrastructure
   see :ref:`West extending signing <west-extending-signing>` for further
   details.
 
+* Fixed an issue whereby when using ``*_ROOT`` variables with Sysbuild, these
+  were lost for images.
+
+* Enhanced ``zephyr_get`` CMake helper function to optionally support merging
+  of scoped variables into a list.
+
+* Added a new CMake helper function for setting/updating sysbuild CMake cache
+  variables: ``sysbuild_cache_set``.
+
 Drivers and Sensors
 *******************
 

--- a/share/sysbuild/cmake/modules/sysbuild_extensions.cmake
+++ b/share/sysbuild/cmake/modules/sysbuild_extensions.cmake
@@ -384,5 +384,70 @@ function(sysbuild_module_call)
       endforeach()
     endif()
   endforeach()
+endfunction()
 
+# Usage:
+#   sysbuild_cache_set(VAR <variable> [APPEND [REMOVE_DUPLICATES]] <value>)
+#
+# This function will set the specified value of the sysbuild cache variable in
+# the CMakeCache.txt file which can then be accessed by images.
+# `VAR` specifies the variable name to set/update.
+#
+# The result will be returned in `<variable>`.
+#
+# Example use:
+#   sysbuild_cache_set(VAR ATTRIBUTES APPEND REMOVE_DUPLICATES battery)
+#     Will add the item `battery` to the `ATTRIBUTES` variable as a new element
+#     in the list in the CMakeCache and remove any duplicates from the list.
+#
+# <variable>:        Name of variable in CMake cache.
+# APPEND:            If specified then will append the supplied data to the
+#                    existing value as a list.
+# REMOVE_DUPLICATES: If specified then remove duplicate entries contained
+#                    within the list before saving to the cache.
+# <value>:           Value to set/update.
+function(sysbuild_cache_set)
+  cmake_parse_arguments(VARS "APPEND;REMOVE_DUPLICATES" "VAR" "" ${ARGN})
+
+  zephyr_check_arguments_required(sysbuild_cache_set VARS VAR)
+
+  if(NOT VARS_UNPARSED_ARGUMENTS AND VARS_APPEND)
+    # Nothing to append so do nothing
+    return()
+  elseif(VARS_REMOVE_DUPLICATES AND NOT VARS_APPEND)
+    message(FATAL_ERROR
+            "sysbuild_set(VAR <var> APPEND REMOVE_DUPLICATES ...) missing required APPEND option")
+  endif()
+
+  get_property(var_type CACHE ${VARS_VAR} PROPERTY TYPE)
+  get_property(var_help CACHE ${VARS_VAR} PROPERTY HELPSTRING)
+
+  # If the variable type is not set, use UNINITIALIZED which will not apply any
+  # specific formatting.
+  if(NOT var_type)
+    set(var_type "UNINITIALIZED")
+  endif()
+
+  if(VARS_APPEND)
+    set(var_new "$CACHE{${VARS_VAR}}")
+
+    # Search for these exact items in the existing value and prevent adding
+    # them if they are already present which avoids issues with double addition
+    # when cmake is reran.
+    string(FIND "$CACHE{${VARS_VAR}}" "${VARS_UNPARSED_ARGUMENTS}" index)
+
+    if(NOT ${index} EQUAL -1)
+      return()
+    endif()
+
+    list(APPEND var_new "${VARS_UNPARSED_ARGUMENTS}")
+
+    if(VARS_REMOVE_DUPLICATES)
+      list(REMOVE_DUPLICATES var_new)
+    endif()
+  else()
+    set(var_new "${VARS_UNPARSED_ARGUMENTS}")
+  endif()
+
+  set(${VARS_VAR} "${var_new}" CACHE "${var_type}" "${var_help}" FORCE)
 endfunction()

